diff --git a/ext/repo_rpmdb.c b/ext/repo_rpmdb.c
index b66d29f..83aee33 100644
--- a/ext/repo_rpmdb.c
+++ b/ext/repo_rpmdb.c
@@ -3027,7 +3027,11 @@ pubkey2solvable(Solvable *s, Repodata *data, char *pubkey)
   repodata_set_str(data, s - s->repo->pool->solvables, PUBKEY_KEYID, keyid);
   if (dig->pubkey.userid)
     setutf8string(data, s - s->repo->pool->solvables, SOLVABLE_SUMMARY, dig->pubkey.userid);
+#ifndef RPM5
   (void)pgpFreeDig(dig);
+#else
+  (void)pgpDigFree(dig);
+#endif
   sat_free((void *)pkts);
   return 1;
 }
